File: elixir/lib/2018/02_inventory_management_system/aoc201802.ex
defmodule AOC2018.Day02 do
  @moduledoc """
  Advent of Code 2018, day 2: Inventory Management System

  With apologies and thanks: https://www.twitch.tv/videos/346878818
  """
  require AOC

  def parse(puzzle_input) do
    puzzle_input |> String.split("\n") |> Enum.map(&String.to_charlist/1)
  end

  def part1(input) do
    input
    |> count_twos_and_threes()
    |> then(fn {twos, threes} -> twos * threes end)
  end

  defp count_characters(box_id), do: count_characters(box_id, %{})
  defp count_characters([], count_acc), do: count_acc

  defp count_characters([char | rest], count_acc) do
    count_characters(rest, Map.update(count_acc, char, 1, &(&1 + 1)))
  end

  defp any_twos_and_threes(counts) do
    counts
    |> Enum.reduce({0, 0}, fn
      {_char, 2}, {_twos, threes} -> {1, threes}
      {_char, 3}, {twos, _threes} -> {twos, 1}
      _, acc -> acc
    end)
  end

  defp count_twos_and_threes(box_ids) do
    box_ids
    |> Enum.reduce({0, 0}, fn box_id, {twos_acc, threes_acc} ->
      {twos, threes} =
        box_id
        |> count_characters()
        |> any_twos_and_threes()

      {twos_acc + twos, threes_acc + threes}
    end)
  end

  def part2(input) do
    input |> find_similar() |> Enum.reverse() |> List.to_string()
  end

  defp find_similar([box_id | box_ids]) do
    box_ids |> Enum.find_value(&find_similar(&1, box_id, [], 0)) || find_similar(box_ids)
  end

  defp find_similar([head | tail_1], [head | tail_2], same, num_diffs) do
    find_similar(tail_1, tail_2, [head | same], num_diffs)
  end

  defp find_similar([_ | tail_1], [_ | tail_2], same, num_diffs) do
    find_similar(tail_1, tail_2, same, num_diffs + 1)
  end

  defp find_similar([], [], same, 1), do: same
  defp find_similar([], [], _same, _), do: nil

  def main(args) do
    Enum.map(args, fn path -> AOC.solve(path, &parse/1, &part1/1, &part2/1) end)
  end
end

File: python/2018/02_inventory_management_system/aoc201802.py
"""AoC 2, 2018: Inventory Management System"""

# Standard library imports
import collections
import itertools
import pathlib
import sys


def parse(puzzle_input):
    """Parse input"""
    return puzzle_input.split("\n")


def part1(data):
    """Solve part 1"""
    twos, threes = zip(*[has_counts(box_id, 2, 3) for box_id in data])
    return sum(twos) * sum(threes)


def part2(data):
    """Solve part 2"""
    return next(
        match
        for first, second in itertools.product(data, repeat=2)
        if (match := almost_equal(first, second))
    )


def has_counts(box_id, *counts):
    """Check if box ID has the given counts of letters

    ## Examples:

    >>> has_counts("geirarne", 2, 3, 5)
    (True, False, False)
    >>> has_counts("geirarnehjelle", 2, 3)
    (True, False)
    """
    box_id_counts = set(collections.Counter(box_id).values())
    return tuple(count in box_id_counts for count in counts)


def almost_equal(first, second):
    """Check if two strings differ in exactly one position.

    Return all matching characters if almost equal.

    ## Examples

    >>> almost_equal("geir", "arne")
    False
    >>> almost_equal("geirarne", "geirarne")
    False
    >>> almost_equal("arne", "anne")
    'ane'
    >>> almost_equal("oooooAooo", "oooooSooo")
    'oooooooo'
    """
    matches = []
    has_missed = False
    for char1, char2 in zip(first, second):
        if char1 == char2:
            matches.append(char1)
        elif not has_missed:
            has_missed = True
        else:
            return False
    return "".join(matches) if has_missed else False


def solve(puzzle_input):
    """Solve the puzzle for the given input"""
    data = parse(puzzle_input)
    yield part1(data)
    yield part2(data)


if __name__ == "__main__":
    for path in sys.argv[1:]:
        print(f"\n{path}:")
        solutions = solve(puzzle_input=pathlib.Path(path).read_text().strip())
        print("\n".join(str(solution) for solution in solutions))
